### ISSUE TITLE
[BUG]cmdmod.exec_code_all use  runas execute the  shell script reteurn code 126

### ISSUE THREAD

**Issue by vekergu** [saltstack/salt#60072]
**Description**
salt   "*" cmd.exec_code_all   bash   "echo  good"  runas=macuser  

cmdmod.exec_code_all use  runas execute the  shell script ，return No Execution Permission，return code 126.

salt use  salt.utils.files.mkstemp  create codefile  is  root  user，but  use  runas    args，other user can not  Execution Permission to run  this codefile.

**Comment by sagetherage** [comment]
@vekergu can you please give us your set up and 
**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 
please?

**Comment by vekergu** [comment]
Salt Version:
          Salt: 3003
 
Dependency Versions:
          cffi: 1.9.1
      cherrypy: unknown
      dateutil: Not Installed
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 2.11.1
       libgit2: Not Installed
      M2Crypto: 0.35.2
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.14
      pycrypto: Not Installed
  pycryptodome: Not Installed
        pygit2: Not Installed
        Python: 3.6.8 (default, Nov 16 2020, 16:55:22)
  python-gnupg: Not Installed
        PyYAML: 3.13
         PyZMQ: 17.0.0
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.1.4
 
System Versions:
          dist: centos 7 Core
        locale: UTF-8
       machine: x86_64
       release: 3.10.0-1160.21.1.el7.x86_64
        system: Linux
       version: CentOS Linux 7 Core


[root@codedemo01 rabbitmq]# salt "192.168.5.5" cmd.exec_code_all bash "echo good" runas=rabbitmq
192.168.5.5:
    ----------
    pid:
        119470
    retcode:
        126
    stderr:
        bash: /tmp/__salt.tmp.pss3czes: Permission denied
    stdout:
ERROR: Minions returned with non-zero exit code

**Comment by s0undt3ch** [comment]
Confirmed, here's a fix with a test
```patch
commit ad3fcaf16c7f1fa8c9d8c80931762c06255b3524 (HEAD -> refs/heads/hotfix/cmdrun-code-all)
Author: Pedro Algarvio <pedro@algarvio.me>
Date:   Wed May 12 07:03:27 2021 +0100

    Ensure proper access to the created temporary file
    
    Fixes #60072

diff --git a/salt/modules/cmdmod.py b/salt/modules/cmdmod.py
index 80d374508b..4e12bbfa58 100644
--- a/salt/modules/cmdmod.py
+++ b/salt/modules/cmdmod.py
@@ -3195,8 +3195,25 @@ def exec_code_all(lang, code, cwd=None, args=None, **kwargs):
     elif isinstance(args, list):
         cmd += args
 
+    def _cleanup_tempfile(path):
+        try:
+            __salt__["file.remove"](path)
+        except (SaltInvocationError, CommandExecutionError) as exc:
+            log.error(
+                "cmd.exec_code_all: Unable to clean tempfile '%s': %s",
+                path,
+                exc,
+                exc_info_on_loglevel=logging.DEBUG,
+            )
+
+    runas = kwargs.get("runas")
+    if runas is not None:
+        if not salt.utils.platform.is_windows():
+            os.chmod(codefile, 320)
+            os.chown(codefile, __salt__["file.user_to_uid"](runas), -1)
+
     ret = run_all(cmd, cwd=cwd, python_shell=False, **kwargs)
-    os.remove(codefile)
+    _cleanup_tempfile(codefile)
     return ret
 
 
diff --git a/tests/pytests/integration/modules/test_cmdmod.py b/tests/pytests/integration/modules/test_cmdmod.py
new file mode 100644
index 0000000000..9e1411bfa5
--- /dev/null
+++ b/tests/pytests/integration/modules/test_cmdmod.py
@@ -0,0 +1,15 @@
+import pytest
+
+
+@pytest.fixture(scope="module")
+def non_root_account():
+    with pytest.helpers.create_account() as account:
+        yield account
+
+
+@pytest.mark.skip_if_not_root
+def test_exec_code_all(salt_call_cli, non_root_account):
+    ret = salt_call_cli.run(
+        "cmd.exec_code_all", "bash", "echo good", runas=non_root_account.username
+    )
+    assert ret.exitcode == 0
```

**Comment by s0undt3ch** [comment]
The PR fixing this https://github.com/saltstack/salt/pull/60173
